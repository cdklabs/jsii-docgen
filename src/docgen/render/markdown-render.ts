import { ApiReferenceSchema, AssemblyMetadataSchema, ClassSchema, ConstructSchema, EnumMemberSchema, EnumSchema, InitializerSchema, InterfaceSchema, JsiiEntity, MethodSchema, ParameterSchema, PropertySchema, StructSchema, TypeSchema } from '../schema';
import { Language } from '../transpile/transpile';
import { MarkdownDocument } from './markdown-doc';

export interface MarkdownFormattingOptions {
  /**
   * How jsii entity IDs should be formatted into anchors. This should be
   * customized in conjunction with `linkFormatter`.
   *
   * @default - use the full id
   */
  readonly anchorFormatter?: (type: JsiiEntity) => string;

  /**
   * How should links to entities be rendered. For example, if a class or a
   * property is referenced within a description or table.
   *
   * @default - '<a href="#{type.id}">{type.displayName}</a>'
   */
  readonly linkFormatter?: (type: JsiiEntity) => string;

  /**
   * How type signatures should be formatted, including those made of nested
   * types (like `Map<string, Bucket>`).
   *
   * @default - HTML code block with type references linked
   * according to `linkFormatter`
   */
  readonly typeFormatter?: (type: TypeSchema, linkFormatter: (type: JsiiEntity) => string) => string;
}

export interface MarkdownRendererOptions extends MarkdownFormattingOptions {
  /**
   * Name of the jsii assembly/package.
   */
  readonly packageName: string;

  /**
   * Version of the jsii assembly/package.
   */
  readonly packageVersion: string;

  /**
   * Name of the submodule this type is from within the jsii assembly (if any).
   */
  readonly submodule?: string;

  /**
   * Language the documentation is rendered for.
   */
  readonly language: Language;
}

export class MarkdownRenderer {
  private readonly anchorFormatter: (type: JsiiEntity) => string;
  private readonly linkFormatter: (type: JsiiEntity) => string;
  private readonly typeFormatter: (type: TypeSchema, linkFormatter: (type: JsiiEntity) => string) => string;
  private readonly metadata: AssemblyMetadataSchema;
  private readonly language: Language;

  constructor(options: MarkdownRendererOptions) {
    this.anchorFormatter = options.anchorFormatter ?? defaultAnchorFormatter;
    this.linkFormatter = options.linkFormatter ?? defaultLinkFormatter;
    this.typeFormatter = options.typeFormatter ?? defaultTypeFormatter;
    this.language = options.language;
    this.metadata = {
      packageName: options.packageName,
      packageVersion: options.packageVersion,
      submodule: options.submodule,
    };
  }

  public visitApiReference(
    apiRef: ApiReferenceSchema,
  ): MarkdownDocument {
    const md = new MarkdownDocument({ header: { title: 'API Reference' }, id: 'api-reference' });
    md.section(this.visitConstructs(apiRef.constructs));
    md.section(this.visitStructs(apiRef.structs));
    md.section(this.visitClasses(apiRef.classes));
    md.section(this.visitInterfaces(apiRef.interfaces));
    md.section(this.visitEnums(apiRef.enums));
    return md;
  }

  public visitConstructs(
    constructs: ConstructSchema[],
  ): MarkdownDocument {
    if (constructs.length === 0) {
      return MarkdownDocument.EMPTY;
    }

    const md = new MarkdownDocument({ header: { title: 'Constructs' } });
    for (const construct of constructs) {
      md.section(this.visitConstruct(construct));
    }
    return md;
  }

  public visitStructs(
    structs: StructSchema[],
  ): MarkdownDocument {
    if (structs.length === 0) {
      return MarkdownDocument.EMPTY;
    }

    const md = new MarkdownDocument({ header: { title: 'Structs' } });
    for (const struct of structs) {
      md.section(this.visitStruct(struct));
    }
    return md;
  }

  public visitClasses(
    classes: ClassSchema[],
  ): MarkdownDocument {
    if (classes.length === 0) {
      return MarkdownDocument.EMPTY;
    }

    const md = new MarkdownDocument({ header: { title: 'Classes' } });
    for (const klass of classes) {
      md.section(this.visitClass(klass));
    }
    return md;
  }

  public visitInterfaces(
    ifaces: InterfaceSchema[],
  ): MarkdownDocument {
    if (ifaces.length === 0) {
      return MarkdownDocument.EMPTY;
    }

    const md = new MarkdownDocument({ header: { title: 'Protocols' } });
    for (const iface of ifaces) {
      md.section(this.visitInterface(iface));
    }
    return md;
  }

  public visitEnums(
    enums: EnumSchema[],
  ): MarkdownDocument {
    if (enums.length === 0) {
      return MarkdownDocument.EMPTY;
    }

    const md = new MarkdownDocument({ header: { title: 'Enums' } });
    for (const enu of enums) {
      md.section(this.visitEnum(enu));
    }
    return md;
  }

  public visitConstruct(
    construct: ConstructSchema,
  ): MarkdownDocument {
    return this.visitClass(construct);
  }

  public visitStruct(
    struct: StructSchema,
  ): MarkdownDocument {
    const md = new MarkdownDocument({
      id: this.anchorFormatter({
        id: struct.id,
        displayName: struct.displayName,
        fqn: struct.fqn,
        ...this.metadata,
      }),
      header: { title: struct.displayName },
    });

    if (struct.docs) {
      md.docs(struct.docs);
    }

    const initializer = new MarkdownDocument({
      id: `${struct.id}.Initializer`,
      header: { title: 'Initializer' },
    });

    if (struct.usage) {
      initializer.code(this.language.toString(), struct.usage);
    }

    md.section(initializer);
    md.section(this.visitProperties(struct.properties));
    return md;
  }

  public visitClass(
    klass: ClassSchema,
  ): MarkdownDocument {
    const md = new MarkdownDocument({
      id: this.anchorFormatter({
        id: klass.id,
        displayName: klass.displayName,
        fqn: klass.fqn,
        ...this.metadata,
      }),
      header: { title: klass.displayName },
    });

    if (klass.interfaces.length > 0) {
      const ifaces = [];
      for (const iface of klass.interfaces) {
        ifaces.push(this.linkFormatter(iface));
      }
      md.bullet(`${MarkdownDocument.italic('Implements:')} ${ifaces.join(', ')}`);
      md.lines('');
    }

    if (klass.docs) {
      md.docs(klass.docs);
    }

    if (klass.initializer) {
      md.section(this.visitInitializer(klass.initializer));
    }

    md.section(this.visitInstanceMethods(klass.instanceMethods));
    md.section(this.visitStaticFunctions(klass.staticMethods));
    md.section(this.visitProperties(klass.properties));
    md.section(this.visitConstants(klass.constants));
    return md;
  }

  public visitInterface(
    iface: InterfaceSchema,
  ): MarkdownDocument {
    const md = new MarkdownDocument({
      id: this.anchorFormatter({
        id: iface.id,
        displayName: iface.displayName,
        fqn: iface.fqn,
        ...this.metadata,
      }),
      header: { title: iface.displayName },
    });

    if (iface.interfaces.length > 0) {
      const bases = [];
      for (const base of iface.interfaces) {
        bases.push(this.linkFormatter(base));
      }
      md.bullet(`${MarkdownDocument.italic('Extends:')} ${bases.join(', ')}`);
      md.lines('');
    }

    if (iface.implementations.length > 0) {
      const impls = [];
      for (const impl of iface.implementations) {
        impls.push(this.linkFormatter(impl));
      }
      md.bullet(`${MarkdownDocument.italic('Implemented By:')} ${impls.join(', ')}`);
      md.lines('');
    }

    if (iface.docs) {
      md.docs(iface.docs);
    }

    md.section(this.visitInstanceMethods(iface.instanceMethods));
    md.section(this.visitProperties(iface.properties));
    return md;
  }

  public visitEnum(
    enu: EnumSchema,
  ): MarkdownDocument {
    const md = new MarkdownDocument({
      id: this.anchorFormatter({
        id: enu.id,
        displayName: enu.displayName,
        fqn: enu.fqn,
        ...this.metadata,
      }),
      header: { title: enu.displayName },
    });

    md.table(this.createTable(enu.members));
    md.split();

    if (enu.docs) {
      md.docs(enu.docs);
    }

    for (const em of enu.members) {
      md.section(this.visitEnumMember(em));
    }

    return md;
  }

  public visitProperties(
    properties: PropertySchema[],
  ): MarkdownDocument {
    if (properties.length === 0) {
      return MarkdownDocument.EMPTY;
    }

    const md = new MarkdownDocument({ header: { title: 'Properties' } });

    md.table(this.createTableWithTypes(properties));
    md.split();

    for (const prop of properties) {
      md.section(this.visitProperty(prop));
    }

    return md;
  }

  public visitInitializer(
    init: InitializerSchema,
  ): MarkdownDocument {
    const md = new MarkdownDocument({
      id: this.anchorFormatter({
        id: init.id,
        displayName: init.displayName,
        fqn: init.fqn,
        ...this.metadata,
      }),
      header: {
        title: 'Initializers',
      },
    });

    if (init.usage) {
      md.code(this.language.toString(), init.usage);
    }

    md.table(this.createTableWithTypes(init.parameters));
    md.split();

    for (const param of init.parameters) {
      md.section(this.visitParameter(param));
    }

    return md;
  }

  public visitInstanceMethods(
    methods: MethodSchema[],
  ): MarkdownDocument {
    if (methods.length === 0) {
      return MarkdownDocument.EMPTY;
    }

    const md = new MarkdownDocument({ header: { title: 'Methods' } });

    md.table(this.createTable(methods));
    md.split();

    for (const method of methods) {
      md.section(this.visitInstanceMethod(method));
    }
    return md;
  }

  public visitStaticFunctions(
    methods: MethodSchema[],
  ) {
    if (methods.length === 0) {
      return MarkdownDocument.EMPTY;
    }

    const md = new MarkdownDocument({ header: { title: 'Static Functions' } });

    md.table(this.createTable(methods));
    md.split();

    for (const method of methods) {
      md.section(this.visitStaticFunction(method));
    }
    return md;
  }

  public visitConstants(
    constants: PropertySchema[],
  ): MarkdownDocument {
    if (constants.length === 0) {
      return MarkdownDocument.EMPTY;
    }

    const md = new MarkdownDocument({ header: { title: 'Constants' } });

    md.table(this.createTableWithTypes(constants));
    md.split();

    for (const con of constants) {
      md.section(this.visitConstant(con));
    }

    return md;
  }

  public visitEnumMember(
    em: EnumMemberSchema,
  ): MarkdownDocument {
    const md = new MarkdownDocument({
      id: this.anchorFormatter({
        id: em.id,
        displayName: em.displayName,
        fqn: em.fqn,
        ...this.metadata,
      }),
      header: {
        title: em.displayName,
        pre: true,
        strike: em.docs.deprecated,
      },
    });

    if (em.docs.deprecated) {
      md.bullet(
        `${MarkdownDocument.italic('Deprecated:')} ${em.docs.deprecationReason}`,
      );
      md.lines('');
    }

    if (em.docs) {
      md.docs(em.docs);
    }

    md.split();
    md.lines('');

    return md;
  }

  public visitProperty(
    prop: PropertySchema,
  ): MarkdownDocument {
    const optionality = prop.optional
      ? 'Optional'
      : 'Required';

    const md = new MarkdownDocument({
      id: this.anchorFormatter({
        id: prop.id,
        displayName: prop.displayName,
        fqn: prop.fqn,
        ...this.metadata,
      }),
      header: {
        title: prop.displayName,
        sup: optionality,
        pre: true,
        strike: prop.docs.deprecated,
      },
    });

    if (prop.docs.deprecated) {
      md.bullet(
        `${MarkdownDocument.italic('Deprecated:')} ${prop.docs.deprecationReason}`,
      );
      md.lines('');
    }

    if (prop.usage) {
      md.code(this.language.toString(), prop.usage);
    }

    const metadata: Record<string, string> = {
      Type: this.typeFormatter(prop.type, this.linkFormatter),
    };

    if (prop.default) {
      metadata.Default = MarkdownDocument.sanitize(prop.default);
    }

    for (const [key, value] of Object.entries(metadata)) {
      md.bullet(`${MarkdownDocument.italic(`${key}:`)} ${value}`);
    }
    md.lines('');

    if (prop.docs) {
      md.docs(prop.docs);
    }

    md.split();

    return md;
  }

  public visitParameter(
    parameter: ParameterSchema,
  ): MarkdownDocument {
    const optionality = parameter.optional ? 'Optional' : 'Required';

    const md = new MarkdownDocument({
      id: this.anchorFormatter({
        id: parameter.id,
        displayName: parameter.displayName,
        fqn: parameter.fqn,
        ...this.metadata,
      }),
      header: {
        title: parameter.displayName,
        sup: optionality,
        pre: true,
        strike: parameter.docs.deprecated,
      },
    });

    if (parameter.docs.deprecated) {
      md.bullet(
        `${MarkdownDocument.italic('Deprecated:')} ${parameter.docs.deprecationReason}`,
      );
      md.lines('');
    }


    const metadata: any = {
      Type: this.typeFormatter(parameter.type, this.linkFormatter),
    };

    if (parameter.default) {
      metadata.Default = MarkdownDocument.sanitize(parameter.default);
    }

    for (const [key, value] of Object.entries(metadata)) {
      md.bullet(`${MarkdownDocument.italic(`${key}:`)} ${value}`);
    }
    md.lines('');

    if (parameter.docs) {
      md.docs(parameter.docs);
    }

    md.split();

    return md;
  }

  public visitInstanceMethod(
    method: MethodSchema,
  ): MarkdownDocument {
    const md = new MarkdownDocument({
      id: this.anchorFormatter({
        id: method.id,
        displayName: method.displayName,
        fqn: method.fqn,
        ...this.metadata,
      }),
      header: {
        title: method.displayName,
        pre: true,
        strike: method.docs.deprecated,
      },
    });

    if (method.usage) {
      md.code(this.language.toString(), method.usage);
    }

    for (const param of method.parameters) {
      md.section(this.visitParameter(param));
    }

    return md;
  }

  public visitStaticFunction(
    method: MethodSchema,
  ) {
    const md = new MarkdownDocument({
      id: this.anchorFormatter({
        id: method.id,
        displayName: method.displayName,
        fqn: method.fqn,
        ...this.metadata,
      }),
      header: {
        title: method.displayName,
        pre: true,
        strike: method.docs.deprecated,
      },
    });

    if (method.usage) {
      md.code(this.language.toString(), method.usage);
    }

    for (const param of method.parameters) {
      md.section(this.visitParameter(param));
    }

    return md;
  }

  public visitConstant(
    constant: PropertySchema,
  ): MarkdownDocument {
    return this.visitProperty(constant);
  }

  private createTable(items: SimpleTableItem[]): string[][] {
    const tableRows: string[][] = [];
    tableRows.push(['Name', 'Description'].map(MarkdownDocument.bold));

    for (const item of items) {
      const link = MarkdownDocument.pre(this.linkFormatter({
        fqn: item.fqn,
        displayName: item.displayName,
        id: item.id,
        ...this.metadata,
      }));
      const description = item.docs?.summary && item.docs?.summary.length > 0
        ? item.docs?.summary
        : MarkdownDocument.italic('No description.');
      tableRows.push([link, description]);
    }

    return tableRows;
  }

  private createTableWithTypes(items: TypeTableItem[]): string[][] {
    const tableRows: string[][] = [];
    tableRows.push(['Name', 'Type', 'Description'].map(MarkdownDocument.bold));

    for (const item of items) {
      const link = MarkdownDocument.pre(this.linkFormatter({
        fqn: item.fqn,
        displayName: item.displayName,
        id: item.id,
        ...this.metadata,
      }));
      const type = MarkdownDocument.pre(this.typeFormatter(item.type, this.linkFormatter));
      const description = item.docs?.summary && item.docs?.summary.length > 0
        ? item.docs?.summary
        : MarkdownDocument.italic('No description.');
      tableRows.push([link, type, description]);
    }

    return tableRows;
  }
}

interface SimpleTableItem {
  readonly fqn: string;
  readonly displayName: string;
  readonly id: string;
  readonly docs?: { summary?: string };
}

interface TypeTableItem extends SimpleTableItem {
  readonly type: TypeSchema;
}

export const defaultAnchorFormatter = (type: JsiiEntity) => {
  return type.id;
};

export const defaultLinkFormatter = (type: JsiiEntity) => {
  return `<a href="#${type.id}">${type.displayName}</a>`;
};

function isJsiiType(value: any): value is JsiiEntity {
  return (
    value !== null
    && typeof value === 'object'
    && value?.fqn
    && value?.id
    && value?.displayName
  );
}

export const defaultTypeFormatter = (
  type: TypeSchema,
  linkFormatter: (type: JsiiEntity) => string,
): string => {
  let result = type.formattingPattern;
  const typeRefs = [];
  for (const typeRef of type.types ?? []) {
    if (isJsiiType(typeRef)) {
      typeRefs.push(linkFormatter(typeRef));
    } else {
      typeRefs.push(defaultTypeFormatter(typeRef, linkFormatter));
    }
  }

  // substitute referred types into the original string
  const placeholderMatcher = /\%/g;
  for (const typeRef of typeRefs) {
    const matches = placeholderMatcher.exec(result);
    if (!matches) {
      // it's possible the number of %'s doesn't match the number of types provided
      // e.g. csharp unions are currently rendered to `{ name: 'object', types: [type1, type2] }`
      continue;
    }
    const insertionIdx: number = matches.index;
    result = result.substring(0, insertionIdx) + typeRef + result.substring(insertionIdx + 1);
  }

  return result;
};