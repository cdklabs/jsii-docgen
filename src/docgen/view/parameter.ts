import * as reflect from 'jsii-reflect';
import { extractDocs, ParameterSchema } from '../schema';
import { Transpile, TranspiledCallable, TranspiledParameter } from '../transpile/transpile';

export class Parameter {
  private readonly transpiledParam: TranspiledParameter;
  private readonly transpiledCallable: TranspiledCallable;
  constructor(
    private readonly transpile: Transpile,
    private readonly parameter: reflect.Parameter,
  ) {
    this.transpiledParam = transpile.parameter(parameter);
    this.transpiledCallable = transpile.callable(parameter.method);
  }

  public toJson(): ParameterSchema {
    let typeschema = this.transpiledParam.typeReference.toJson();
    // If the parameter is variadic, then we have to change the formatting pattern of the type schema.
    // For example, if the formatting pattern for a string type is 'string', we'd make it variadic by
    // changing it to '...string[]' in Typescript - each language has its own variadic syntax which is defined
    // in the `variadicOf()` function.
    //
    // Another example: let's say we have a Python function that takes a single number parameter like so: 'number: int'
    // To make this variadic, we'd change it to '*number: int' -- here, the formatting pattern needs to be updated to prepend
    // the '*' character to indicate the parameter is variadic.
    if (this.transpiledParam.variadic) {
      typeschema = {
        formattingPattern: this.transpile.variadicOf(typeschema.formattingPattern),
        types: typeschema.types,
      };
    }
    const isInitializer = this.parameter.method.kind === reflect.MemberKind.Initializer;
    const methodName = isInitializer ? 'Initializer' : this.transpiledCallable.name;
    const methodId = isInitializer ? 'Initializer' : this.parameter.method.name;
    return {
      fqn: `${this.transpiledParam.parentType.fqn}.${methodName}.parameter.${this.transpiledParam.name}`,
      displayName: this.transpiledParam.name,
      id: `${this.parameter.parentType.fqn}.${methodId}.parameter.${this.parameter.name}`,
      optional: this.transpiledParam.optional === true ? true : undefined, // to save space
      default: this.parameter.spec.docs?.default,
      type: typeschema,
      variadic: this.transpiledParam.variadic,
      docs: extractDocs(this.parameter.docs),
    };
  }
}
