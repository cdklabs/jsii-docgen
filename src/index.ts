export { Documentation } from './docgen/view/documentation';
export { Language } from './docgen/transpile/transpile';
export { UnsupportedLanguageError } from './docgen/transpile/transpile';
export { TranspiledType } from './docgen/transpile/transpile';
export * from './errors';
