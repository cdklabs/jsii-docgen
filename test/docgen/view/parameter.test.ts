import * as reflect from 'jsii-reflect';
import { Language } from '../../../src';
import { MarkdownRenderer } from '../../../src/docgen/render/markdown-render';
import { TypeScriptTranspile } from '../../../src/docgen/transpile/typescript';
import { LANGUAGE_SPECIFIC } from '../../../src/docgen/view/documentation';
import { Parameter } from '../../../src/docgen/view/parameter';
import { Assemblies } from '../assemblies';

const assembly: reflect.Assembly = Assemblies.instance.withoutSubmodules;
const variadicTestAssembly: reflect.Assembly = Assemblies.instance.withVariadicParameter;

const metadata = {
  packageName: assembly.name,
  packageVersion: assembly.version,
};

const findParameter = (): reflect.Parameter => {
  for (const klass of assembly.system.classes) {
    for (const method of klass.ownMethods) {
      if (method.parameters.length > 0) {
        return method.parameters[0];
      }
    }
  }
  throw new Error('Assembly does not contain a parameter');
};

test.each(Language.values())('%s snapshot', (language) => {
  const { transpile } = LANGUAGE_SPECIFIC[language.toString()];
  const markdown = new MarkdownRenderer({ language, ...metadata });
  const param = new Parameter(transpile, findParameter()).toJson();
  expect(param).toMatchSnapshot();
  expect(markdown.visitParameter(param).render()).toMatchSnapshot();
});

test('newlines in "defaults" are removed', () => {
  const transpile = new TypeScriptTranspile();
  const reflectParameter = findParameter();
  reflectParameter.spec.docs = { default: 'default option\nwith a newline' };
  const docgenParameter = new Parameter(transpile, reflectParameter).toJson();
  const renderer = new MarkdownRenderer({ language: Language.TYPESCRIPT, ...metadata });
  const markdown = renderer.visitParameter(docgenParameter).render();
  expect(markdown).toMatchSnapshot();
  expect(markdown).toContain('default option with a newline');
});

test('newlines in any other than "defaults" are not removed', () => {
  const transpile = new TypeScriptTranspile();
  const reflectParameter = findParameter();
  reflectParameter.spec.docs = { remarks: 'remarks\nwith a newline' };
  const docgenParameter = new Parameter(transpile, reflectParameter).toJson();
  const renderer = new MarkdownRenderer({ language: Language.TYPESCRIPT, ...metadata });
  const markdown = renderer.visitParameter(docgenParameter).render();
  expect(markdown).toContain('remarks\nwith a newline');
});

// Tests variadic parameter in initializer
const findInitializerVariadicParameter = (): reflect.Parameter => {
  for (const klass of variadicTestAssembly.classes) {
    if (klass.initializer) {
      for (const param of klass.initializer.parameters) {
        if (param.variadic) {
          return param;
        }
      }
    }
  }
  throw new Error('Assembly does not contain a variadic parameter');
};

test.each(Language.values())('%s snapshot variadic in initializer', (language) => {
  const { transpile } = LANGUAGE_SPECIFIC[language.toString()];
  const renderer = new MarkdownRenderer({ language, ...metadata });
  const param = new Parameter(transpile, findInitializerVariadicParameter()).toJson();
  const markdown = renderer.visitParameter(param).render();
  expect(markdown).toMatchSnapshot();
});

// Test variadic param in Instance Method
const findInstanceMethodVariadicParameter = (): reflect.Parameter => {
  for (const klass of variadicTestAssembly.classes) {
    for (const method of klass.ownMethods) {
      if (!method.static) {
        for (const param of method.parameters) {
          if (param.variadic) {
            return param;
          }
        }
      }
    }
  }
  throw new Error('Assembly does not contain an instance method with a variadic parameter');
};
test.each(Language.values())('%s snapshot variadic in instance method', (language) => {
  const { transpile } = LANGUAGE_SPECIFIC[language.toString()];
  const renderer = new MarkdownRenderer({ language, ...metadata });
  const param = new Parameter(transpile, findInstanceMethodVariadicParameter()).toJson();
  const markdown = renderer.visitParameter(param).render();
  expect(markdown).toMatchSnapshot();
});

// Test variadic param in Static Method
const findStaticMethodVariadicParameter = (): reflect.Parameter => {
  for (const klass of variadicTestAssembly.classes) {
    for (const method of klass.ownMethods) {
      if (method.static) {
        for (const param of method.parameters) {
          if (param.variadic) {
            return param;
          }
        }
      }
    }
  }
  throw new Error('Assembly does not contain a static method with a variadic parameter');
};
test.each(Language.values())('%s snapshot variadic in static function', (language) => {
  const { transpile } = LANGUAGE_SPECIFIC[language.toString()];
  const renderer = new MarkdownRenderer({ language, ...metadata });
  const param = new Parameter(transpile, findStaticMethodVariadicParameter()).toJson();
  const markdown = renderer.visitParameter(param).render();
  expect(markdown).toMatchSnapshot();
});
