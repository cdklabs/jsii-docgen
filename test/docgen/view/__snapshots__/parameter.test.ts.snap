// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`csharp snapshot 1`] = `
{
  "default": undefined,
  "displayName": "Grantee",
  "docs": {},
  "fqn": "Amazon.CDK.AWS.ECR.AuthorizationToken.GrantRead.parameter.Grantee",
  "id": "@aws-cdk/aws-ecr.AuthorizationToken.grantRead.parameter.grantee",
  "optional": undefined,
  "type": {
    "formattingPattern": "%",
    "types": [
      {
        "displayName": "IGrantable",
        "fqn": "Amazon.CDK.AWS.IAM.IGrantable",
        "id": "@aws-cdk/aws-iam.IGrantable",
        "packageName": "@aws-cdk/aws-iam",
        "packageVersion": "99.99.99",
      },
    ],
  },
  "variadic": false,
}
`;

exports[`csharp snapshot 2`] = `
" \`Grantee\`<sup>Required</sup> <a name="Grantee" id="@aws-cdk/aws-ecr.AuthorizationToken.grantRead.parameter.grantee"></a>

- *Type:* Amazon.CDK.AWS.IAM.IGrantable

---
"
`;

exports[`csharp snapshot variadic in initializer 1`] = `
" \`VariadicProps\`<sup>Required</sup> <a name="VariadicProps" id="variadic-jsii-example.MyVariadicClass.Initializer.parameter.variadicProps"></a>

- *Type:* params double[]

---
"
`;

exports[`csharp snapshot variadic in instance method 1`] = `
" \`VariadicInput\`<sup>Required</sup> <a name="VariadicInput" id="variadic-jsii-example.MyVariadicClass.myFunction.parameter.variadicInput"></a>

- *Type:* params double[]

---
"
`;

exports[`csharp snapshot variadic in static function 1`] = `
" \`StaticVariadicInput\`<sup>Required</sup> <a name="StaticVariadicInput" id="variadic-jsii-example.MyVariadicClass.myStaticFunction.parameter.staticVariadicInput"></a>

- *Type:* params double[]

---
"
`;

exports[`go snapshot 1`] = `
{
  "default": undefined,
  "displayName": "grantee",
  "docs": {},
  "fqn": "github.com/aws/aws-cdk-go/awscdkawsecr.AuthorizationToken.GrantRead.parameter.grantee",
  "id": "@aws-cdk/aws-ecr.AuthorizationToken.grantRead.parameter.grantee",
  "optional": undefined,
  "type": {
    "formattingPattern": "%",
    "types": [
      {
        "displayName": "IGrantable",
        "fqn": "github.com/aws/aws-cdk-go/awscdk/awsiam/awscdkawsiam.IGrantable",
        "id": "@aws-cdk/aws-iam.IGrantable",
        "packageName": "@aws-cdk/aws-iam",
        "packageVersion": "99.99.99",
      },
    ],
  },
  "variadic": false,
}
`;

exports[`go snapshot 2`] = `
" \`grantee\`<sup>Required</sup> <a name="grantee" id="@aws-cdk/aws-ecr.AuthorizationToken.grantRead.parameter.grantee"></a>

- *Type:* github.com/aws/aws-cdk-go/awscdk/awsiam/awscdkawsiam.IGrantable

---
"
`;

exports[`go snapshot variadic in initializer 1`] = `
" \`variadicProps\`<sup>Required</sup> <a name="variadicProps" id="variadic-jsii-example.MyVariadicClass.Initializer.parameter.variadicProps"></a>

- *Type:* ...*f64

---
"
`;

exports[`go snapshot variadic in instance method 1`] = `
" \`variadicInput\`<sup>Required</sup> <a name="variadicInput" id="variadic-jsii-example.MyVariadicClass.myFunction.parameter.variadicInput"></a>

- *Type:* ...*f64

---
"
`;

exports[`go snapshot variadic in static function 1`] = `
" \`staticVariadicInput\`<sup>Required</sup> <a name="staticVariadicInput" id="variadic-jsii-example.MyVariadicClass.myStaticFunction.parameter.staticVariadicInput"></a>

- *Type:* ...*f64

---
"
`;

exports[`java snapshot 1`] = `
{
  "default": undefined,
  "displayName": "grantee",
  "docs": {},
  "fqn": "software.amazon.awscdk.services.ecr.AuthorizationToken.grantRead.parameter.grantee",
  "id": "@aws-cdk/aws-ecr.AuthorizationToken.grantRead.parameter.grantee",
  "optional": undefined,
  "type": {
    "formattingPattern": "%",
    "types": [
      {
        "displayName": "IGrantable",
        "fqn": "software.amazon.awscdk.services.iam.IGrantable",
        "id": "@aws-cdk/aws-iam.IGrantable",
        "packageName": "@aws-cdk/aws-iam",
        "packageVersion": "99.99.99",
      },
    ],
  },
  "variadic": false,
}
`;

exports[`java snapshot 2`] = `
" \`grantee\`<sup>Required</sup> <a name="grantee" id="@aws-cdk/aws-ecr.AuthorizationToken.grantRead.parameter.grantee"></a>

- *Type:* software.amazon.awscdk.services.iam.IGrantable

---
"
`;

exports[`java snapshot variadic in initializer 1`] = `
" \`variadicProps\`<sup>Required</sup> <a name="variadicProps" id="variadic-jsii-example.MyVariadicClass.Initializer.parameter.variadicProps"></a>

- *Type:* java.lang.Number...

---
"
`;

exports[`java snapshot variadic in instance method 1`] = `
" \`variadicInput\`<sup>Required</sup> <a name="variadicInput" id="variadic-jsii-example.MyVariadicClass.myFunction.parameter.variadicInput"></a>

- *Type:* java.lang.Number...

---
"
`;

exports[`java snapshot variadic in static function 1`] = `
" \`staticVariadicInput\`<sup>Required</sup> <a name="staticVariadicInput" id="variadic-jsii-example.MyVariadicClass.myStaticFunction.parameter.staticVariadicInput"></a>

- *Type:* java.lang.Number...

---
"
`;

exports[`newlines in "defaults" are removed 1`] = `
" \`grantee\`<sup>Required</sup> <a name="grantee" id="@aws-cdk/aws-ecr.AuthorizationToken.grantRead.parameter.grantee"></a>

- *Type:* @aws-cdk/aws-iam.IGrantable
- *Default:* default option with a newline

---
"
`;

exports[`python snapshot 1`] = `
{
  "default": undefined,
  "displayName": "grantee",
  "docs": {},
  "fqn": "aws_cdk.aws_ecr.AuthorizationToken.grant_read.parameter.grantee",
  "id": "@aws-cdk/aws-ecr.AuthorizationToken.grantRead.parameter.grantee",
  "optional": undefined,
  "type": {
    "formattingPattern": "%",
    "types": [
      {
        "displayName": "IGrantable",
        "fqn": "aws_cdk.aws_iam.IGrantable",
        "id": "@aws-cdk/aws-iam.IGrantable",
        "packageName": "@aws-cdk/aws-iam",
        "packageVersion": "99.99.99",
      },
    ],
  },
  "variadic": false,
}
`;

exports[`python snapshot 2`] = `
" \`grantee\`<sup>Required</sup> <a name="grantee" id="@aws-cdk/aws-ecr.AuthorizationToken.grantRead.parameter.grantee"></a>

- *Type:* aws_cdk.aws_iam.IGrantable

---
"
`;

exports[`python snapshot variadic in initializer 1`] = `
" \`variadic_props\`<sup>Required</sup> <a name="variadic_props" id="variadic-jsii-example.MyVariadicClass.Initializer.parameter.variadicProps"></a>

- *Type:* *typing.Union[int, float]

---
"
`;

exports[`python snapshot variadic in instance method 1`] = `
" \`variadic_input\`<sup>Required</sup> <a name="variadic_input" id="variadic-jsii-example.MyVariadicClass.myFunction.parameter.variadicInput"></a>

- *Type:* *typing.Union[int, float]

---
"
`;

exports[`python snapshot variadic in static function 1`] = `
" \`static_variadic_input\`<sup>Required</sup> <a name="static_variadic_input" id="variadic-jsii-example.MyVariadicClass.myStaticFunction.parameter.staticVariadicInput"></a>

- *Type:* *typing.Union[int, float]

---
"
`;

exports[`typescript snapshot 1`] = `
{
  "default": undefined,
  "displayName": "grantee",
  "docs": {},
  "fqn": "@aws-cdk/aws-ecr.AuthorizationToken.grantRead.parameter.grantee",
  "id": "@aws-cdk/aws-ecr.AuthorizationToken.grantRead.parameter.grantee",
  "optional": undefined,
  "type": {
    "formattingPattern": "%",
    "types": [
      {
        "displayName": "IGrantable",
        "fqn": "@aws-cdk/aws-iam.IGrantable",
        "id": "@aws-cdk/aws-iam.IGrantable",
        "packageName": "@aws-cdk/aws-iam",
        "packageVersion": "99.99.99",
      },
    ],
  },
  "variadic": false,
}
`;

exports[`typescript snapshot 2`] = `
" \`grantee\`<sup>Required</sup> <a name="grantee" id="@aws-cdk/aws-ecr.AuthorizationToken.grantRead.parameter.grantee"></a>

- *Type:* @aws-cdk/aws-iam.IGrantable

---
"
`;

exports[`typescript snapshot variadic in initializer 1`] = `
" \`variadicProps\`<sup>Required</sup> <a name="variadicProps" id="variadic-jsii-example.MyVariadicClass.Initializer.parameter.variadicProps"></a>

- *Type:* ...number[]

---
"
`;

exports[`typescript snapshot variadic in instance method 1`] = `
" \`variadicInput\`<sup>Required</sup> <a name="variadicInput" id="variadic-jsii-example.MyVariadicClass.myFunction.parameter.variadicInput"></a>

- *Type:* ...number[]

---
"
`;

exports[`typescript snapshot variadic in static function 1`] = `
" \`staticVariadicInput\`<sup>Required</sup> <a name="staticVariadicInput" id="variadic-jsii-example.MyVariadicClass.myStaticFunction.parameter.staticVariadicInput"></a>

- *Type:* ...number[]

---
"
`;
